### ISSUE TITLE
Git checkout failure on QT submodule (Windows)

### ISSUE THREAD

**Issue by XenHat** [telegramdesktop/tdesktop#2125]
Hello, I am trying to build Telegram for Windows Desktop and I am encountering a problem with the build instructions:

The current QT preparation steps mentions this:
````
perl init-repository --module-subset=qtbase,qtimageformats
git checkout v5.6.0
````
This doesn't seem to work properly for me and I am not really familiar with git in advanced scenarios so I am unsure of what to do.
Here is a more comprehensive log:

````bash
> perl init-repository --module-subset=qtbase,qtimageformats
# No problem here
> git checkout v5.6.0
# ~~~~~~~~~~~~~^ This creates a DETACHED head:
M       qtbase
M       qtimageformats
Note: checking out 'v5.6.0'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 84a1676... Update .gitmodules for automated release src package creation

TBuild\Libraries\qt5_6_0> cd qtimageformats && git checkout v5.6.0 && cd ..
error: Your local changes to the following files would be overwritten by checkout:
        src/3rdparty/libwebp.pri
        src/plugins/imageformats/dds/dds.pro
        src/plugins/imageformats/icns/icns.pro
        src/plugins/imageformats/icns/qicnshandler.cpp
        src/plugins/imageformats/jp2/jp2.pro
        src/plugins/imageformats/mng/mng.pro
        src/plugins/imageformats/tga/tga.pro
        src/plugins/imageformats/tiff/qtiffhandler.cpp
        src/plugins/imageformats/tiff/tiff.pro
        src/plugins/imageformats/wbmp/wbmp.pro
        src/plugins/imageformats/webp/webp.pro
        tests/auto/tiff/tst_qtiff.cpp
Please, commit your changes or stash them before you can switch branches.
Aborting
````

What am I supposed to do?

### Configuration
**Operating system:**
Windows 10 1156 (OS Build 10586) x86_64

**Comment by XenHat** [comment]
For the record, committing the changes and moving forward after checkout does not work:

```bash
E:\devel\general\TBuild\Libraries\qt5_6_0>cd qtbase && git apply ../../../tdesktop/Telegram/Patches/qtbase_5_6_0.diff && cd ..
error: patch failed: mkspecs/common/msvc-desktop.conf:30
error: mkspecs/common/msvc-desktop.conf: patch does not apply
error: patch failed: src/corelib/tools/qunicodetables.cpp:6227
error: src/corelib/tools/qunicodetables.cpp: patch does not apply
error: patch failed: src/gui/kernel/qplatformdialoghelper.h:386
error: src/gui/kernel/qplatformdialoghelper.h: patch does not apply
error: patch failed: src/gui/painting/qpaintengine_p.h:80
error: src/gui/painting/qpaintengine_p.h: patch does not apply
error: patch failed: src/gui/text/qtextlayout.h:196
error: src/gui/text/qtextlayout.h: patch does not apply
error: patch failed: src/platformsupport/fontdatabases/basic/qbasicfontdatabase.cpp:172
error: src/platformsupport/fontdatabases/basic/qbasicfontdatabase.cpp: patch does not apply
error: patch failed: src/plugins/platforminputcontexts/platforminputcontexts.pro:1
error: src/plugins/platforminputcontexts/platforminputcontexts.pro: patch does not apply
error: patch failed: src/plugins/platforms/cocoa/qcocoaapplicationdelegate.mm:210
error: src/plugins/platforms/cocoa/qcocoaapplicationdelegate.mm: patch does not apply
error: patch failed: src/plugins/platforms/cocoa/qcocoabackingstore.mm:38
error: src/plugins/platforms/cocoa/qcocoabackingstore.mm: patch does not apply
error: patch failed: src/plugins/platforms/cocoa/qcocoakeymapper.mm:384
error: src/plugins/platforms/cocoa/qcocoakeymapper.mm: patch does not apply
error: patch failed: src/plugins/platforms/cocoa/qcocoasystemtrayicon.mm:94
error: src/plugins/platforms/cocoa/qcocoasystemtrayicon.mm: patch does not apply
error: patch failed: src/plugins/platforms/windows/qwindowswindow.h:265
error: src/plugins/platforms/windows/qwindowswindow.h: patch does not apply
error: patch failed: src/widgets/dialogs/qfiledialog.h:108
error: src/widgets/dialogs/qfiledialog.h: patch does not apply
error: patch failed: src/widgets/dialogs/qfiledialog_p.h:123
error: src/widgets/dialogs/qfiledialog_p.h: patch does not apply
error: patch failed: src/widgets/widgets/qwidgettextcontrol.cpp:71
error: src/widgets/widgets/qwidgettextcontrol.cpp: patch does not apply
```

**Comment by stek29** [comment]
You should checkout, and only after that apply patch

**Comment by XenHat** [comment]
This is what I'm *trying* to do, the process errors out before I even reach the patch apply step...

**Comment by john-preston** [comment]
You get an error "Your local changes to the following files would be overwritten by checkout". That is strange (if you didn't change anything there yet).

Do you try a fresh clone of git repository? Not some that you used already for some builds or anything?

**Comment by XenHat** [comment]
Fresh clone, attempted three times from scratch.

**Comment by john-preston** [comment]
I tried it right now and everything went fine. This commands one at the time:

git clone git://code.qt.io/qt/qt5.git qt5_6_0
cd qt5_6_0
git checkout 5.6
perl init-repository --module-subset=qtbase,qtimageformats
git checkout v5.6.0
cd qtimageformats && git checkout v5.6.0 && cd ..
cd qtbase && git checkout v5.6.0 && cd ..

in a fresh folder. They were launched in "VS2015 x86 Native Tools Command Prompt".

**Comment by XenHat** [comment]
```bash
:\Program Files (x86)\Microsoft Visual Studio 14.0\VC>cd /d E:\devel\general\TBuild

E:\devel\general\TBuild>cd Libraries

E:\devel\general\TBuild\Libraries>git clone git://code.qt.io/qt/qt5.git qt5_6_0git clone git://code.qt.io/qt/qt5.git qt5_6_0

E:\devel\general\TBuild\Libraries>git clone git://code.qt.io/qt/qt5.git qt5_6_0
Cloning into 'qt5_6_0'...
remote: Counting objects: 3674, done.
remote: Compressing objects: 100% (2186/2186), done.
Receiving objects: 100% (3674/3674), 3.94 MiB | 1.21 MiB/s, done.
Resolving deltas:  16% (303/1863)   ed 2816 (delta 1408)
Resolving deltas: 100% (1863/1863), done.
Checking connectivity... done.

E:\devel\general\TBuild\Libraries>cd qt5_6_0

E:\devel\general\TBuild\Libraries\qt5_6_0>git checkout 5.6
Already on '5.6'
Your branch is up-to-date with 'origin/5.6'.

E:\devel\general\TBuild\Libraries\qt5_6_0>perl init-repository --module-subset=qtbase,qtimageformats --quiet
Cloning into 'qtbase'...
remote: Counting objects: 413285, done.
remote: Compressing objects: 100% (94834/94834), done.
Receiving objects: 100% (413285/413285), 201.80 MiB | 3.07 MiB/s,

Resolving deltas: 100% (327665/327665), done.
Checking connectivity... done.
Checking out files: 100% (21312/21312), done.
Cloning into 'qtimageformats'...
remote: Counting objects: 3155, done.
remote: Compressing objects: 100% (2047/2047), done.
remote: Total 3155 (delta 1478), reused 2137 (delta 1027)R
Receiving objects: 100% (3155/3155), 4.40 MiB | 2.01 MiB/s, done.
Resolving deltas: 100% (1478/1478), done.
Checking connectivity... done.
Submodule path 'qtbase': checked out 'fd70978693bd92761e9989bc1c76bf83c1e8c987'

E:\devel\general\TBuild\Libraries\qt5_6_0>git checkout v5.6.0
M       qtbase
M       qtimageformats
Note: checking out 'v5.6.0'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 84a1676... Update .gitmodules for automated release src package creation

E:\devel\general\TBuild\Libraries\qt5_6_0>cd qtimageformats

E:\devel\general\TBuild\Libraries\qt5_6_0\qtimageformats>git checkout v5.6.0
error: Your local changes to the following files would be overwritten by checkout:
        .qmake.conf
        src/3rdparty/libwebp.pri
        src/plugins/imageformats/dds/dds.pro
        src/plugins/imageformats/icns/icns.pro
        src/plugins/imageformats/icns/qicnshandler.cpp
        src/plugins/imageformats/jp2/jp2.pro
        src/plugins/imageformats/mng/mng.pro
        src/plugins/imageformats/tga/tga.pro
        src/plugins/imageformats/tiff/qtiffhandler.cpp
        src/plugins/imageformats/tiff/tiff.pro
        src/plugins/imageformats/wbmp/wbmp.pro
        src/plugins/imageformats/webp/webp.pro
        tests/auto/tiff/tst_qtiff.cpp
Please, commit your changes or stash them before you can switch branches.
Aborting
```

is my dev machine possessed?

**Comment by john-preston** [comment]
Looks like we have different git clients? Mine has strange many lines in the first command:
```bash
T:\TBuild\tmp>git clone git://code.qt.io/qt/qt5.git qt5_6_0
Cloning into 'qt5_6_0'...
remote: Counting objects: 3674, done.
remote: Compressing objects:  15% (328/2186)   R
remote: Compressing objects:  43% (940/2186)   Receiving objects:   1% (37/3674)remote: Compressing objects:  46% (1006/2186)   R
remote: Compressing objects:  66% (1443/2186)    eceiving objects:   4% (147/367remote: Compressing objects:  73% (1596/2186)   Receiving objects:   5% (184/367remote: Compressing objects:  74% (1618/2186)   R
remote: Compressing objects:  75% (1640/2186)   R
remote: Compressing objects:  92% (2012/2186)   R
remote: Compressing objects:  99% (2165/2186)   R
remote: Compressing objects: 100% (2186/2186)   R
remote: Compressing objects: 100% (2186/2186), done.R
/3674)
Receiving objects: 100% (3674/3674), 3.94 MiB | 2.43 MiB/s, done.
emote: Total 3674 (delta 1866), reused 2816 (delta 1408)Resolving deltas:  24% (Resolving deltas:  31% (581/1866)
Resolving deltas: 100% (1866/1866), done.
Checking connectivity... done.
```
Perhaps some settings in git (core.autocrlf?) make changes to those files automatically? If I check for changes after first git checkout v5.6.0 I get this:
```bash
T:\TBuild\tmp\qt5_6_0>cd qtimageformats

T:\TBuild\tmp\qt5_6_0\qtimageformats>git diff

T:\TBuild\tmp\qt5_6_0\qtimageformats>git status -u
On branch 5.6
Your branch is up-to-date with 'origin/5.6'.

nothing to commit, working directory clean

T:\TBuild\tmp\qt5_6_0\qtimageformats>
```

**Comment by john-preston** [comment]
You can run "git checkout ." before "git checkout v5.6.0" in "qtimageformats", perhaps it will work that way. I'm not sure the patch will apply if you have such problems with own Qt checkouts %)

**Comment by XenHat** [comment]
I am using Atlassian's Sourcetree. I will look into the configuration to
see if I can reset some options, and failing that, will use standalone git.

**Comment by XenHat** [comment]
I seem to have resolved it myself by "switching" to the `v5.6.0` branch manually with `git checkout -b "v5.6.0"`.
So strange.

Full log:
```bash
E:\devel\general\TBuild\Libraries\qt5_6_0\qtimageformats>cd qtimageformats

E:\devel\general\TBuild\Libraries\qt5_6_0\qtimageformats>git status -u
HEAD detached at v5.6.0
nothing to commit, working directory clean

E:\devel\general\TBuild\Libraries\qt5_6_0\qtimageformats>git checkout -b "v5.6.0"
Switched to a new branch 'v5.6.0'

E:\devel\general\TBuild\Libraries\qt5_6_0\qtimageformats>git status -u
On branch v5.6.0
nothing to commit, working directory clean

E:\devel\general\TBuild\Libraries\qt5_6_0\qtimageformats>git checkout v5.6.0
warning: refname 'v5.6.0' is ambiguous.
Already on 'v5.6.0'

E:\devel\general\TBuild\Libraries\qt5_6_0\qtbase>git apply ../../../tdesktop/Telegram/Patches/qtbase_5_6_0.diff

E:\devel\general\TBuild\Libraries\qt5_6_0\qtbase>
```

**Comment by XenHat** [comment]
For the record, I cleaned my Git config and re-installed it, picking "checkout as-is, commit as-is" and added the whole thing to the path (through the installer) and it seems to have resolved my "unstaged changes" issue.
